File: JavaScript_Implementation/FruitsTree/Tree.mjs
import { Fruit, ovalShaped, Berry, tinyFruit, Citrus, Apple, Avocado, Blackberries, Elderberries, Lemon, Orange, Tangerine } from './Fruit.mjs';
class Tree {
    constructor() {
        this.root = null;
    }

    insert(weight, type) {
        var newFruit;
        switch (type) {
            case "ovalShaped":
                newFruit = Object.create(ovalShaped.prototype);
                break;
            case "TinyFruit":
                newFruit = Object.create(tinyFruit.prototype);
                break;
            case "Citrus":
                newFruit = Object.create(Citrus.prototype);
                break;
            case "Berry":
                newFruit = Object.create(Berry.prototype);
                break;
            case "Apple":
                newFruit = Object.create(Apple.prototype);
                break;
            case "Avocado":
                newFruit = Object.create(Avocado.prototype);
                break;
            case "Blackberry":
                newFruit = Object.create(Blackberries.prototype);
                break;
            case "Elderberry":
                newFruit = Object.create(Elderberries.prototype);
                break;
            case "Lemon":
                newFruit = Object.create(Lemon.prototype);
                break;
            case "Orange":
                newFruit = Object.create(Orange.prototype);
                break;
            case "Tangerine":
                newFruit = Object.create(Tangerine.prototype);
            default:
                newFruit = Object.create(Fruit.prototype);
                break;
        }
        newFruit.weight = weight;
        newFruit.left = null;
        newFruit.right = null;
        if (this.root === null)
            this.root = newFruit;
        else
            this.insertHelper.call(this, this.root, newFruit);
    }

    insertHelper(node, newFruit) {
        if (newFruit.getWeight() < node.getWeight()) {
            //if left is null insert node here
            if (node.getLeft() === null) {
                node.left = newFruit;
                node.left.parent = node;
            } else {
                this.insertHelper.call(this, node.left, newFruit);
            }
        } else if (newFruit.getWeight() > node.getWeight()) {
            // if right is null insert node here
            if (node.getRight() === null) {
                node.right = newFruit;
                node.right.parent = node;
            } else {
                this.insertHelper.call(this, node.right, newFruit);
            }
        } else {
            newFruit.parent = node.parent;
            newFruit.left = node.left;
            newFruit.right = node.right;
            if (node.left !== null)
                node.left.parent = newFruit;
            if (node.right !== null)
                node.right.parent = newFruit;

            if (node.parent !== null) {
                if (node.parent.left === node)
                    node.parent.left = newFruit;
                else
                    node.parent.right = newFruit;
            } else {
                this.root = newFruit;
            }
        }
    }

    iterate() {
        let list = [];
        this.getList.call(this, this.root, list);
        list.forEach(x => {
            console.log(x.getType() + " " + x.getWeight() + " gm");
        });
    }

    filterByType(type) {
        let list = [];
        let outputList = [];
        this.getList.call(this, this.root, list);
        list.forEach(x => {
            if (x instanceof type) {
                outputList.push(x)
            }
        });
        return outputList
    }

    filterByWeight(weight) {
        let list = [];
        this.getList.call(this, this.root, list);
        list.forEach(x => {
            if (x.getWeight() > weight) {
                console.log(x.getType() + " " + x.getWeight() + " gm");
            }
        });
    }

    magnifyByType(type, weight) {
        let removed = this.filterByType.call(this, type);
        removed.forEach(x => {
            this.remove.call(this, x.getWeight());
        });
        removed.forEach(x => {
            x.weight += weight;
        });
        removed.forEach(x => {
            this.insert.call(this, x.getWeight(), x.getType());
        });
    }

    findHeaviest() {
        return this.findMax.call(this, this.root);
    }

    findLightest() {
        return this.findMin.call(this, this.root);
    }

    getList(node, list) {
        if (node !== null) {
            this.getList.call(this, node.left, list);
            list.push(node)
            this.getList.call(this, node.right, list);
        }
    }

    remove(weight) {
        this.root = this.removeNode.call(this, this.root, weight);
    }

    removeNode(node, weight) {
        // if the root is null then tree is
        // empty
        if (node === null)
            return null;

        // if data to be delete is less than
        // roots data then move to left subtree
        else if (weight < node.getWeight()) {
            node.left = this.removeNode.call(this, node.left, weight);
            return node;
        }

        // if data to be delete is greater than
        // roots data then move to right subtree
        else if (weight > node.getWeight()) {
            node.right = this.removeNode.call(this, node.right, weight);
            return node;
        }

        // if data is similar to the root's data
        // then delete this node
        else {
            // deleting node with no children
            if (node.left === null && node.right === null) {
                node = null;
                return node;
            }

            if (node.left === null) {
                node = node.right;
                return node;
            } else if (node.right === null) {
                node = node.left;
                return node;
            }

            var aux = this.findMin.call(this, node.right);
            if (node.parent !== null) {
                if (node.parent.getLeft === node)
                    node.parent.left = aux;
                else
                    node.parent.right = aux;
            }

            if (aux.parent.left === aux)
                aux.parent.left = null;
            else
                aux.parent.right = null;
            aux.parent = node.parent;
            aux.right = node.right;
            aux.left = node.left;
            if (node.left !== null)
                node.left.parent = aux;
            if (node.right !== null)
                node.right.parent = aux;
            node = aux;
            return node;
        }
    }

    findMin(node) {
        if (node.left === null)
            return node;
        else
            return this.findMin.call(this, node.left);
    }

    findMax(node) {
        if (node.right === null)
            return node;
        else
            return this.findMax.call(this, node.right);
    }
}


var BST = new Tree();
BST.insert(12, "Apple");
BST.insert(13, "Lemon");
BST.insert(5, "Avocado");
BST.insert(8, "Blackberry");

BST.iterate();
console.log("------------------------");
BST.filterByType(Citrus).forEach(x => {
    console.log(x.getWeight() + " " + x.getType());
});
console.log("------------------------");
BST.filterByWeight(12);
console.log("------------------------");
BST.magnifyByType(Apple, 4);
BST.iterate();
console.log("------------------------");
console.log(BST.findLightest().getWeight() + " " + BST.findLightest().getType());
console.log("------------------------");
console.log(BST.findHeaviest().getWeight() + " " + BST.findHeaviest().getType());
console.log("------------------------");
BST.magnifyByType(ovalShaped, 4);
BST.iterate();

File: Cpp_Implementation/FruitsTree/Tree.cpp
#include "Tree.h"

Tree::Tree() {root = nullptr;}

void Tree::insert(int weight) {
    Fruit* newFruit;
/*
    if (type == "OvalShaped")
        OvalShaped* newFruit = reinterpret_cast<OvalShaped*>(newFruit);
    else if (type == "TinyFruit")
        TinyFruit* newFruit = reinterpret_cast<TinyFruit*>(newFruit);
    else if (type == "Berry")
        Berry* newFruit = reinterpret_cast<Berry*>(newFruit);
    else if (type == "Apple")
        Apple* newFruit = reinterpret_cast<Apple*>(newFruit);
    else if (type == "Avocado")
        Avocado newFruit;
    else if (type == "Blackberry")
        Blackberry newFruit;
    else if (type == "Elderberry")
        ElderBerry newFruit;
    else
        Fruit newFruit;
*/
    newFruit->weight = weight;
    newFruit->left = nullptr;
    newFruit->right = nullptr;
    if (root == nullptr)
        root = newFruit;
    else
        insertHelper(root, newFruit);
}

void Tree::insertHelper(Fruit* fruit, Fruit* newFruit) {
    if (newFruit->weight < fruit->weight) {
        if(fruit->left == nullptr){
            fruit->left = newFruit;
            fruit->left->parent = fruit;
        } else {
            insertHelper(fruit->left, newFruit);
        }
    } else if (newFruit->weight > fruit->weight) {
        if(fruit->right == nullptr){
            fruit->right = newFruit;
            fruit->right->parent = fruit;
        } else {
            insertHelper(fruit->right, newFruit);
        }
    } else {
        newFruit->parent = fruit->parent;
        newFruit->left = fruit->left;
        newFruit->right = fruit->right;
        if (fruit->left != nullptr)
            fruit->left->parent = newFruit;
        if (fruit->right != nullptr)
            fruit->right->parent = newFruit;
        if (fruit->parent != nullptr){
            if (fruit->parent->left == fruit)
                fruit->parent->left = newFruit;
            else
                fruit->parent->right = newFruit;
        } else {
            root = newFruit;
        }
    }
}

void Tree::inorderTraversal(Fruit* fruit) {
    if (fruit == nullptr)
        return;
    inorderTraversal(fruit->left);
    cout << fruit->getType() << " " << fruit->weight << "gm" << endl;
    inorderTraversal(fruit->right);
}

void Tree::iterate() {
    inorderTraversal(root);
}

void Tree::filterByType(string type) {
    filterByTypeHelper(root, type);
}

void Tree::filterByTypeHelper(Fruit* fruit, string type) {
    if (fruit == nullptr)
        return;
    filterByTypeHelper(fruit->left, type);
}

void Tree::filterByWeight(int weight) {
    filterByWeightHelper(root, weight);
}

void Tree::filterByWeightHelper(Fruit* fruit, int weight) {

}

void Tree::magnifyByType(string type, int weight) {
    list<int> weightList;
    magnifyByTypeHelper(root, type, weight, weightList);
    list<int> :: iterator it;
    for (it = weightList.begin(); it != weightList.end(); ++it)
        insert(*it);
}

void Tree::magnifyByTypeHelper(Fruit* fruit, string type, int weight, list<int> weightList) {
    if (fruit == nullptr)
        return;
    magnifyByTypeHelper(fruit->left, type, weight, weightList);
//if (fruit->getType())
//is_base_of<Fruit, fruit->getType()>
//Base* bp = dynamic_cast<Base*>(&d1)
}

void Tree::remove(int weight) {
    root = removeHelper(root, weight);
}

Fruit *Tree::removeHelper(Fruit* fruit, int weight) {
    if (fruit == nullptr) {
        return nullptr;
    } else if (weight < fruit->weight) {
        fruit->left = removeHelper(fruit->left, weight);
        return fruit;
    } else if (weight > fruit->weight) {
        fruit->right = removeHelper(fruit->right, weight);
        return fruit;
    } else {
        if (fruit->left == nullptr && fruit->right == nullptr) {
            fruit = nullptr;
            return fruit;
        }
        if (fruit->left == nullptr) {
            fruit = fruit->right;
            return fruit;
        } else if (fruit->right == nullptr) {
            fruit = fruit->left;
            return fruit;
        }

        Fruit* temp = findLightestHelper(fruit->right);
        if (fruit->parent != nullptr) {
            if (fruit->parent->left == fruit)
                fruit->parent->left = temp;
            else
                fruit->parent->right = temp;
        }
        if (temp->parent->left == temp)
            temp->parent->left = nullptr;
        else
            temp->parent->right = nullptr;
        temp->parent = fruit->parent;
        temp->right = fruit->right;
        temp->left = fruit->left;
        if (fruit->left != nullptr)
            fruit->left->parent = temp;
        if (fruit->right != nullptr)
            fruit->right->parent = temp;
        fruit = temp;
        return fruit;

    }
    return nullptr;
}

Fruit* Tree::findLightest() {
    return findLightestHelper(root);
}

Fruit* Tree::findLightestHelper(Fruit* fruit) {
    if (fruit->left == nullptr)
        return fruit;
    return findLightestHelper(fruit->left);
}

Fruit* Tree::findHeaviest() {
    return findHeaviestHelper(root);
}

Fruit* Tree::findHeaviestHelper(Fruit* fruit) {
    if (fruit->right == nullptr)
        return fruit;
    return findHeaviestHelper(fruit->right);
}

